### PR TITLE
PHPStan now processes generics for models.

### LINKED ISSUE #11708: PHPStan should check for generics if possible.
### Mautic Version

5.0.x series

### PHP version

8.0.x

### What browsers are you seeing the problem on?

Not relevant

### What happened?

After merging #11587 PHPStan no longer checks for generics. https://github.com/mautic/mautic/pull/11587/files#diff-0361f0c81f363476ddc6f44ab36fcbe66ee685d5f4c2a46b054924591544b766R8

### How can we reproduce this issue?

Step 1: Run PHPStan.

### Relevant log output

_No response_

### Code of Conduct

- [X] I confirm that I have read and agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by biozshock**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [ x ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ x ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #11708 

#### Description:
After merging https://github.com/mautic/mautic/pull/11587 PHPStan no longer checks for generics. https://github.com/mautic/mautic/pull/11587/files#diff-0361f0c81f363476ddc6f44ab36fcbe66ee685d5f4c2a46b054924591544b766R8

#### Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Run composer phpstan

### CODE REVIEW

**Comment by volha-pivavarchyk** [comment]
PHPStan works well.

image

image


@biozshock Is this enough to check?

File: app/bundles/AssetBundle/Controller/AjaxController.php [hunk 1]
```diff
@@ -19,9 +20,11 @@ class AjaxController extends CommonAjaxController
      */
     protected function categoryListAction(Request $request)
     {
-        $filter    = InputHelper::clean($request->query->get('filter'));
-        $results   = $this->getModel('asset')->getLookupResults('category', $filter, 10);
-        $dataArray = [];
+        /** @var AssetModel $assetModel */
```
Comment (escopecz) [inline app/bundles/AssetBundle/Controller/AjaxController.php]: It would be nice to use `\assert($assetModel instanceof AssetModel)` instead of type hints. See https://github.com/mautic/mautic/pull/11019

But this is still a great improvement!

File: app/bundles/CampaignBundle/Controller/Api/EventLogApiController.php [hunk 2]
```diff
@@ -27,20 +35,26 @@ class EventLogApiController extends CommonApiController
...
+        $this->model                    = $campaignEventLogModel;
+        $this->entityClass              = LeadEventLog::class;
         $this->entityNameOne            = 'event';
         $this->entityNameMulti          = 'events';
         $this->parentChildrenLevelDepth = 1;
         $this->serializerGroups         = [
             'campaignList',
             'ipAddressList',
-            'log' => 'campaignEventLogDetails',
+            self::LOG_SERIALIZATION => 'campaignEventLogDetails',
```
Comment (escopecz) [inline app/bundles/CampaignBundle/Controller/Api/EventLogApiController.php:54]: Can you double-check that changing this to `30 => 'campaignEventLogDetails',` is the right thing to do? What was wrong with `'log' => 'campaignEventLogDetails',`?

File: app/bundles/CampaignBundle/Model/EventLogModel.php [hunk 3]
```diff
@@ -32,14 +36,15 @@ public function __construct(
...
-     */
-    public function getRepository()
+    public function getRepository(): LeadEventLogRepository
     {
-        return $this->em->getRepository('MauticCampaignBundle:LeadEventLog');
+        $result = $this->em->getRepository(LeadEventLog::class);
+
+        if (!$result instanceof LeadEventLogRepository) {
+            throw new \RuntimeException('Wrong repository given.');
+        }
```
Comment (escopecz) [inline app/bundles/CampaignBundle/Model/EventLogModel.php]: Checks in this PR like this one could be also replaced by 
```suggestion
        \assert($result instanceof LeadEventLogRepository);
```

File: app/bundles/ConfigBundle/Model/SysinfoModel.php [hunk 4]
```diff
@@ -5,11 +5,12 @@
 use Doctrine\DBAL\Connection;
 use Mautic\CoreBundle\Helper\CoreParametersHelper;
 use Mautic\CoreBundle\Helper\PathsHelper;
+use Mautic\CoreBundle\Model\MauticModelInterface;
 use Mautic\InstallBundle\Configurator\Step\CheckStep;
 use Mautic\InstallBundle\Install\InstallService;
 use Symfony\Contracts\Translation\TranslatorInterface;
 
-class SysinfoModel
+class SysinfoModel implements MauticModelInterface
```
Comment (escopecz) [inline app/bundles/ConfigBundle/Model/SysinfoModel.php]: This doesn't seem to be necessary. Or is it? I'd like to avoid unnecessary inheritance.

File: app/bundles/CoreBundle/Entity/CommonRepository.php [hunk 5]
```diff
@@ -1468,12 +1473,14 @@ protected function buildWhereClause($q, array $args)
                 $queryParameters[$param] = $ids;
             } else {
                 $queryExpression->add(
-                    $q->expr()->in($this->getTableAlias().'.id', $ids)
+                    $q->expr()->in($this->getTableAlias().'.id', array_map(static function (int $id): string {
+                        return (string) $id;
+                    }, $ids))
```
Comment (escopecz) [inline app/bundles/CoreBundle/Entity/CommonRepository.php]: I'm looking at the 2 ways how to do the same thing in the `if` and `else` part. It looks silly that the IDs are retyped to int and right after that to string :)

We should be careful as retyping to int was also meant as query param sanitization. So wherever we change int to string (like below) we should ensure that we use prepared statements.

File: app/bundles/FormBundle/Controller/Api/FormApiController.php [hunk 6]
```diff
@@ -30,14 +44,6 @@ public function initialize(ControllerEvent $event)
         parent::initialize($event);
     }
 
-    /**
-     * {@inheritdoc}
-     */
-    protected function preSerializeEntity(&$entity, $action = 'view')
```
Comment (escopecz) [inline app/bundles/FormBundle/Controller/Api/FormApiController.php:36]: Why was this removed? I worry about changing the behavior in this PR

File: app/bundles/PageBundle/Controller/Api/PageApiController.php [hunk 7]
```diff
@@ -43,12 +55,4 @@ public function getEntitiesAction()
 
         return parent::getEntitiesAction();
     }
-
-    /**
-     * {@inheritdoc}
-     */
-    protected function preSerializeEntity(&$entity, $action = 'view')
```
Comment (escopecz) [inline app/bundles/PageBundle/Controller/Api/PageApiController.php:50]: Why was this removed? I worry about changing the behavior in this PR

File: phpstan.neon [hunk 8]
```diff
@@ -5,7 +5,7 @@ includes:
 parameters:
 	level: 6
 	reportUnmatchedIgnoredErrors: false
-	checkGenericClassInNonGenericObjectType: false
+	#checkGenericClassInNonGenericObjectType: false
```
Comment (escopecz) [inline phpstan.neon]: I guess this line can be removed :)

**Review by escopecz — CHANGES_REQUESTED** [review]
Amazing changes!

I noticed some places I worry about and some places that could be improved. 

For the further refactoring please consider making it in smaller chunks. It is hard not to overlook some issue in this diff.

**Comment by biozshock** [comment]
> Amazing changes!
> 
> I noticed some places I worry about and some places that could be improved.
> 
> For the further refactoring please consider making it in smaller chunks. It is hard not to overlook some issue in this diff.

Unfortunately smaller chunks were not possible: need to add a generic annotation to each model and repository, and then fix all places where PHPStan was complaining, which resulted in this PR.

File: app/bundles/CampaignBundle/Controller/Api/EventLogApiController.php (hunk 2 above)
Comment (biozshock) [reply app/bundles/CampaignBundle/Controller/Api/EventLogApiController.php:54]: Serializer expects groups to be `array<int, string>` (sure it will work with string keys, but the code there is expecting array to be indexed, not associative. This change ensures the `campaignEventLogDetails` group is last in groups list.

File: app/bundles/CampaignBundle/Model/EventLogModel.php (hunk 3 above)
Comment (biozshock) [reply app/bundles/CampaignBundle/Model/EventLogModel.php]: Was thinking about assert, but this change also checks at runtime. Due to the low coverage we can spot these at runtime. But i can change all of these to assert no prob. WDYT?

File: app/bundles/ConfigBundle/Model/SysinfoModel.php (hunk 4 above)
Comment (biozshock) [reply app/bundles/ConfigBundle/Model/SysinfoModel.php]: As the `\Mautic\CoreBundle\Factory\ModelFactory::getModel` now returns a `MauticModelInterface` it's necessary that all models implement this interface. Also in the future it will help with autowiring.

File: app/bundles/FormBundle/Controller/Api/FormApiController.php (hunk 6 above)
Comment (biozshock) [reply app/bundles/FormBundle/Controller/Api/FormApiController.php:36]: It's never used.
EDIT: to be more exact there are no `automaticJs` in the project. It seems that was removed in 2015

File: app/bundles/PageBundle/Controller/Api/PageApiController.php (hunk 7 above)
Comment (biozshock) [reply app/bundles/PageBundle/Controller/Api/PageApiController.php:50]: There is no `url` property in the entity

File: app/bundles/CoreBundle/Entity/CommonRepository.php (hunk 5 above)
Comment (biozshock) [reply app/bundles/CoreBundle/Entity/CommonRepository.php]: See new commit.

File: phpstan.neon (hunk 8 above)
Comment (biozshock) [reply phpstan.neon]: Indeed.

File: app/bundles/CoreBundle/Entity/CommonRepository.php (hunk 5 above)
Comment (biozshock) [reply app/bundles/CoreBundle/Entity/CommonRepository.php]: See new commit.

File: app/bundles/CampaignBundle/Model/EventLogModel.php (hunk 3 above)
Comment (escopecz) [reply app/bundles/CampaignBundle/Model/EventLogModel.php]: I'd recommend to use `/assert()` as it will fail on runtime in the dev and test environments. And it's quite hard (read waste of time) to write tests that will cover the code in these ifs. I'd suggest to replace it. Advantages:
- Shorter, readable code
- Works for IDEs and PHPSTAN
- Fails on test and dev environments
- Better performance in production

File: app/bundles/ConfigBundle/Model/SysinfoModel.php (hunk 4 above)
Comment (escopecz) [reply app/bundles/ConfigBundle/Model/SysinfoModel.php]: I shouldn't have called this class a model probably. It's just a service that have nothing in common with the classic Mautic models that are used work with entities, permissions and so on. I called it a model because it manages the data needed for the system info view. But otherwise doesn't need any special attention that the classic Mautic models need.

File: app/bundles/AssetBundle/Controller/AjaxController.php (hunk 1 above)
Comment (biozshock) [reply app/bundles/AssetBundle/Controller/AjaxController.php]: See latest commits

File: app/bundles/CampaignBundle/Model/EventLogModel.php (hunk 3 above)
Comment (biozshock) [reply app/bundles/CampaignBundle/Model/EventLogModel.php]: See last commits

File: app/bundles/ConfigBundle/Model/SysinfoModel.php (hunk 4 above)
Comment (biozshock) [reply app/bundles/ConfigBundle/Model/SysinfoModel.php]: Removed the interface. Now these models are injected into controllers.

File: app/bundles/CoreBundle/Controller/AbstractFormController.php [hunk 9]
```diff
@@ -240,4 +242,17 @@ protected function refererPostActionVars($vars)
...
+
+    protected function getFormButton(FormInterface $form, array $elements): ClickableInterface
+    {
+        foreach ($elements as $element) {
+            $form = $form->get($element);
+        }
+
+        if (!$form instanceof ClickableInterface) {
+            throw new \RuntimeException('The form item is not ClickableInterface.');
+        }
```
Comment (escopecz) [inline app/bundles/CoreBundle/Controller/AbstractFormController.php]:
```suggestion
        \assert($form instanceof ClickableInterface);
```

File: app/bundles/CoreBundle/Controller/AjaxController.php [hunk 10]
```diff
@@ -278,6 +279,9 @@ protected function togglePublishStatusAction(Request $request)
                         $accessor->setValue($entity, $customToggle, !$accessor->getValue($entity, $customToggle));
                         $model->getRepository()->saveEntity($entity);
                     } else {
+                        if (!$model instanceof FormModel) {
+                            throw new \RuntimeException('The model given is not form model.');
+                        }
```
Comment (escopecz) [inline app/bundles/CoreBundle/Controller/AjaxController.php]:
```suggestion
                        \assert($model instanceof FormModel);
```

File: app/bundles/CoreBundle/Controller/AjaxController.php [hunk 11]
```diff
@@ -336,6 +340,9 @@ protected function unlockEntityAction(Request $request)
             $checkedOut = $entity->getCheckedOutBy();
             if (null !== $entity && !empty($checkedOut) && $checkedOut === $currentUser->getId()) {
                 //entity exists, is checked out, and is checked out by the current user so go ahead and unlock
+                if (!$model instanceof FormModel) {
+                    throw new \RuntimeException('The model given is not form model.');
+                }
```
Comment (escopecz) [inline app/bundles/CoreBundle/Controller/AjaxController.php]:
```suggestion
                \assert($model instanceof FormModel);
```

**Review by escopecz — CHANGES_REQUESTED** [review]
I noticed a couple of places that could be also simplified. I can't wait to get this in!

Just for fun, you can try to regenerate the PHPSTAN baseline in this PR or another. This should remove a lot of errors.
```
php -d memory_limit=4G bin/phpstan analyse --generate-baseline
```

File: app/bundles/CoreBundle/Controller/AbstractFormController.php (hunk 9 above)
Comment (biozshock) [reply app/bundles/CoreBundle/Controller/AbstractFormController.php]: See last commit.

File: app/bundles/CoreBundle/Controller/AjaxController.php (hunk 10 above)
Comment (biozshock) [reply app/bundles/CoreBundle/Controller/AjaxController.php]: See last commit.

File: app/bundles/CoreBundle/Controller/AjaxController.php (hunk 11 above)
Comment (biozshock) [reply app/bundles/CoreBundle/Controller/AjaxController.php]: See last commit.

**Review by escopecz — APPROVED** [review]
Wow, over 5.5% of the PHPSTAN baseline file is gone. That's a huge tech debt improvement. Thank you!
